azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -46,10 +68,111 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
     @staticmethod
     def is_policy_enforcement_enabled():
         """
-        Check whether user has opted into policy enforcement feature.
+        Policy will be enabled if (1) policy file exists at CUSTOM_POLICY_PATH and (2) the conf flag "Debug.EnableExtensionPolicy" is true.
         Caller function should check this before performing any operations.
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
         # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        policy_file_exists = os.path.exists(CUSTOM_POLICY_PATH)
+        return conf.get_extension_policy_enabled() and policy_file_exists
+
+    def __get_policy(self):
+        """
+         Check if custom policy exists at CUSTOM_POLICY_PATH, load JSON object and return as a dict.
+        Return {} if no policy exists.
+        The expected policy format is:
+         {
+            "policyVersion": str,
+            "extensionPolicies": {
+                "allowListedExtensionsOnly": bool,
+                "signatureRequired": bool,
+                "signingPolicy": dict,
+                "extensions": {
+                    "<extensionName1>": {
+                        "signatureRequired": bool,
+                        "signingPolicy": dict,
+                        "runtimePolicy": dict
+                    }
+                }
+            }
+        }
+        """
+        if os.path.exists(CUSTOM_POLICY_PATH):
+            self._log_policy("Custom policy found at {0}. Using custom policy.".format(CUSTOM_POLICY_PATH))
+            with open(CUSTOM_POLICY_PATH, 'r') as f:        # Open file in read-only mode.
+                custom_policy = json.load(f)
+                return custom_policy
+        else:
+            self._log_policy("No custom policy found at {0}. Using default policy.".format(CUSTOM_POLICY_PATH))
+            return {}
+
+
+class ExtensionPolicyEngine(_PolicyEngine):
+
+    def __init__(self, extension_to_check):
```

Comment:
document the expected type of extension_to_check (when i first read this code i thought it was a string)

Reply:
@narrieta since we're initializing on a per goal-state basis, I can remove this parameter in the constructor and take it as a parameter in should_allow() / should_enforce_signature_validation(). Let me know if that makes sense to you.

Reply:
Yes we should remove this parameter since we're initializing on a goal state basis.